csharp/src/Ice/Communicator-ObjectAdapterFactory.cs

```diff
@@ -56,20 +59,24 @@ async Task PerformShutdownAsync(List<ObjectAdapter> adapters)
         /// dispose the runtime and finally exists the application.</summary>
         public async Task WaitForShutdownAsync()
         {
-            Task shutdownTask;
             lock (_mutex)
             {
-                if (_shutdownTask == null)
+                if (_shutdownSemaphore == null)
                 {
                     _waitForShutdownCompletionSource ??= new TaskCompletionSource<object?>();
-                    shutdownTask = _waitForShutdownCompletionSource.Task;
-                }
-                else
-                {
-                    shutdownTask = _shutdownTask;
                 }
             }
-            await shutdownTask.ConfigureAwait(false);
+
+            if (_waitForShutdownCompletionSource != null)
+            {
+                await _waitForShutdownCompletionSource.Task.ConfigureAwait(false);
+            }
+            else
+            {
+                // Wait for the shutdown to complete if shutdown has been initiated.
+                await _shutdownSemaphore.WaitAsync().ConfigureAwait(false);
```

Comment:
Say thread 1 runs ShutdownAsync and creates the semaphore but does not have time to acquire in
```
await _shutdownSemaphore.WaitAsync().ConfigureAwait(false);
```

Then thread 2 runs WaitForShutdownAsync, sees a non-null semaphore and successfully grabs it and releases it before the shutdown is performed.

Maybe we could just always use _waitForShutdownCompletionSource for WaitForShutdownAsync, i.e. create it in both ShutdownAsync and WaitForShutdownAsync if not already created?

Reply:
Good point, I missed that one. We can indeed always create the task completion source for waitForShutdown. I'll do that but to me the initial version based on Yield was simpler and less error prone.